server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -418,9 +426,32 @@ private Repository createRepository(RepositoryMetaData repositoryMetaData, Map<S
             throw new RepositoryException(repositoryMetaData.name(),
                 "repository type [" + repositoryMetaData.type() + "] does not exist");
         }
+
+        List<RepositoryPlugin.RepositoryDecorator> decorators
+            = repositoryDecorators.stream().map(repositoryDecorator -> repositoryDecorator.apply(repositoryMetaData))
+            .filter(Objects::nonNull).collect(Collectors.toList());
+        if (decorators.size() > 1) {
+            throw new RepositoryException(repositoryMetaData.name(),
+                "multiple plugins attempted to decorate repository but this is forbidden: " + decorators);
+        }
+        final RepositoryMetaData repositoryMetaDataAfterDecoration;
+        final Function<Repository, Repository> repositoryWrapper;
+        if (decorators.size() == 1) {
+            Settings.Builder settings = Settings.builder().put(repositoryMetaData.settings());
+            for (final Setting consumedSetting : decorators.get(0).getConsumedSettings()) {
+                settings.remove(consumedSetting.getKey());
```

Comment:
I guess technically we could do without the setting removal for now, we're not doing any setting validation anyway? Seems like currently it just makes things more complicated to no benefit?

Reply:
either is fine with me.